src/deploy/functions/backend.ts

```diff
@@ -0,0 +1,356 @@
+import * as proto from "../../gcp/proto";
+import * as gcf from "../../gcp/cloudfunctions";
+import * as cloudscheduler from "../../gcp/cloudscheduler";
+import * as utils from "../../utils";
+import { FirebaseError } from "../../error";
+import { Context } from "./args";
+
+export interface ScheduleRetryConfig {
+  retryCount?: number;
+  maxRetryDuration?: proto.Duration;
+  minBackoffDuration?: proto.Duration;
+  maxBackoffDuration?: proto.Duration;
+  maxDoublings?: number;
+}
+
+export interface PubSubSpec {
+  id: string;
+  project: string;
+
+  // What we're actually planning to invoke with this topic
+  targetService: FunctionSpec;
+}
+
+export interface ScheduleSpec {
+  id: string;
+  project: string;
+  // Note: schedule is missing in the existingBackend because we
+  // don't actually spend the API call looking up the schedule;
+  // we just infer it from function labels.
+  schedule?: string;
+  timeZone?: string;
+  retryConfig?: ScheduleRetryConfig;
+  transport: "pubsub" | "https";
+
+  // What we're actually planning to invoke with this schedule
+  targetService: FunctionSpec;
+}
+
+export interface HttpsTrigger {
+  httpsOnly: boolean;
+}
+
+export type EventFilterKey = "resource";
+export interface EventTrigger {
+  eventType: string;
+
+  // In v1 API, only "resource" is set.
+  // In v2 API, "resource" will be hoisted up as
+  // a top-level "pubsubTopic" resource for the
+  // google.cloud.pubsub.topic.v1.messagePublished event.
+  eventFilters: Record<EventFilterKey | string, string>;
+
+  retry: boolean;
+
+  // API v2 only. Defaults to "global".
+  region?: string;
+  serviceAccountEmail?: string;
+}
+
+export function isEventTrigger(trigger: HttpsTrigger | EventTrigger): trigger is EventTrigger {
+  return "eventType" in trigger;
+}
+
+export type VpcEgressSettings = "PRIVATE_RANGES_ONLY" | "ALL_TRAFFIC";
+export type IngressSettings = "ALLOW_ALL" | "ALLOW_INTERNAL_ONLY" | "ALLOW_INTERNAL_AND_GCLB";
+export type MemoryOptions = 128 | 256 | 512 | 1024 | 2048 | 4096;
+export type Runtime = "nodejs6" | "nodejs8" | "nodejs10" | "nodejs12" | "nodejs14";
```

Comment:
`nodejs6` and `nodejs8` are not supported runtimes anymore...can we detect this at the point we would otherwise cast to it instead of supporting it here?

Reply:
I wrote the code but am not sure I like it. The problem is that GCF can return a node6 or node8 function so `fromGCFv1Function` has to force-cast the Runtime field. I currently warn, but I'm not sure I like having to manually subvert the type system.